include/flang/Optimizer/Dialect/FIRType.h

```diff
@@ -0,0 +1,357 @@
+  KindTy getFKind() const;
+};
+
+class RealType : public mlir::Type::TypeBase<RealType, mlir::Type,
+                                             detail::RealTypeStorage>,
+                 public IntrinsicTypeMixin<RealType, TypeKind::FIR_REAL> {
+public:
+  using Base::Base;
+  static RealType get(mlir::MLIRContext *ctxt, KindTy kind);
+  KindTy getFKind() const;
+};
+
+// FIR support types
+
+class BoxType
+    : public mlir::Type::TypeBase<BoxType, mlir::Type, detail::BoxTypeStorage> {
+public:
+  using Base::Base;
+  static BoxType get(mlir::Type eleTy, mlir::AffineMapAttr map = {});
+  static bool kindof(unsigned kind) { return kind == TypeKind::FIR_BOX; }
+  mlir::Type getEleTy() const;
+  mlir::AffineMapAttr getLayoutMap() const;
+
+  static mlir::LogicalResult
+  verifyConstructionInvariants(const mlir::AttributeStorage *, mlir::Type eleTy,
+                               mlir::AffineMapAttr map);
+};
+
+class BoxCharType : public mlir::Type::TypeBase<BoxCharType, mlir::Type,
+                                                detail::BoxCharTypeStorage> {
+public:
+  using Base::Base;
+  static BoxCharType get(mlir::MLIRContext *ctxt, KindTy kind);
+  static bool kindof(unsigned kind) { return kind == TypeKind::FIR_BOXCHAR; }
+  CharacterType getEleTy() const;
+};
+
+class BoxProcType : public mlir::Type::TypeBase<BoxProcType, mlir::Type,
+                                                detail::BoxProcTypeStorage> {
+public:
+  using Base::Base;
+  static BoxProcType get(mlir::Type eleTy);
+  static bool kindof(unsigned kind) { return kind == TypeKind::FIR_BOXPROC; }
+  mlir::Type getEleTy() const;
+
+  static mlir::LogicalResult
+  verifyConstructionInvariants(const mlir::AttributeStorage *,
+                               mlir::Type eleTy);
+};
+
+class DimsType : public mlir::Type::TypeBase<DimsType, mlir::Type,
+                                             detail::DimsTypeStorage> {
+public:
+  using Base::Base;
+  static DimsType get(mlir::MLIRContext *ctx, unsigned rank);
+  static bool kindof(unsigned kind) { return kind == TypeKind::FIR_DIMS; }
+
+  /// returns -1 if the rank is unknown
+  int getRank() const;
+};
+
+class FieldType : public mlir::Type::TypeBase<FieldType, mlir::Type,
+                                              detail::FieldTypeStorage> {
+public:
+  using Base::Base;
+  static FieldType get(mlir::MLIRContext *ctxt, KindTy _ = 0);
+  static bool kindof(unsigned kind) { return kind == TypeKind::FIR_FIELD; }
+};
+
+class HeapType : public mlir::Type::TypeBase<HeapType, mlir::Type,
+                                             detail::HeapTypeStorage> {
+public:
+  using Base::Base;
+  static HeapType get(mlir::Type elementType);
+  static bool kindof(unsigned kind) { return kind == TypeKind::FIR_HEAP; }
+
+  mlir::Type getEleTy() const;
+
+  static mlir::LogicalResult
+  verifyConstructionInvariants(const mlir::AttributeStorage *,
+                               mlir::Type eleTy);
+};
+
+class LenType
```

Comment:
Is there a reason why some of these types have comments and others don't?

Reply:
Thanks. I'll document these.